File: Java/Classes/Beneficiary.java
// Beneficiary.java
public class Beneficiary extends User {
    private int noPersons;
    private RequestDonationList receivedList;
    private Requests requestsList;

    public Beneficiary(String name, String phone) {
        super(name, phone);
        this.noPersons = 1;
        this.receivedList = new RequestDonationList();
        this.requestsList = new Requests(this, Organization.getInstance());
    }

    public Beneficiary(String name, String phone, int noPersons) {
        super(name, phone);
        this.noPersons = noPersons;
        this.receivedList = new RequestDonationList();
        this.requestsList = new Requests(this, Organization.getInstance());
    }

    public void add(RequestDonation request) {
        try {
            requestsList.add(request, Organization.getInstance());
            System.out.println("Request added successfully!");
        } catch (Exceptions.EntityNotFoundException e) {
            System.out.println("Error: " + e.getMessage());
            return;
        }
    }

    public void remove(RequestDonation request) {
        requestsList.remove(request);
    }

    public void commit() {
        try {
            requestsList.commit();
        } catch (Exception e) {
            System.out.println("Error committing requests: " + e.getMessage());
        }
    }

    public int getNoPersons() {
        return noPersons;
    }

    public void setNoPersons(int noPersons) {
        this.noPersons = noPersons;
    }

    public RequestDonationList getReceivedList() {
        return receivedList;
    }

    public Requests getRequestsList() {
        return requestsList;
    }
}

File: C++/cpp_files/Beneficiary.cpp
#include "Beneficiary.h"
#include "RequestDonationList.h" // Will be created later
#include "Requests.h" // Will be created later

Beneficiary::Beneficiary(const std::string& name, const std::string& phone, int noPersons)
    : User(name, phone), 
      noPersons(noPersons), 
      receivedList(new RequestDonationList()),
      requestsList(new Requests()) {
    requestsList->setBeneficiary(this);  // Set the beneficiary reference
}

Beneficiary::~Beneficiary() {
    delete receivedList;
    delete requestsList;
}

int Beneficiary::getNoPersons() const {
    return noPersons;
}

void Beneficiary::setNoPersons(int number) {
    noPersons = number;
}

// Wrapper methods for receivedList
void Beneficiary::addReceived(int entityId, double quantity) {
    receivedList->add(entityId, quantity);
}

void Beneficiary::removeReceived(int entityId) {
    receivedList->remove(entityId);
}

void Beneficiary::monitorReceived() const {
    receivedList->monitor();
}

void Beneficiary::resetReceived() {
    receivedList->reset();
}

// Wrapper methods for requestsList
void Beneficiary::addRequest(int entityId, double quantity) {
    requestsList->add(entityId, quantity);
}

void Beneficiary::removeRequest(int entityId) {
    requestsList->remove(entityId);
}

void Beneficiary::commitRequests() {
    requestsList->commit();
}

void Beneficiary::monitorRequests() const {
    requestsList->monitor();
}

void Beneficiary::resetRequests() {
    requestsList->reset();
}